docs/source/core_concepts/index.rst

```diff
@@ -536,6 +536,220 @@ To construct a vector of dense FSAs,  you can either:
+  - `row_indexes = [0, 2, 4, 7, 8]`
+
+**Explanation**:
+  - `values` contains the non-zero entries of the matrix
+  - `values[0] = 10` belongs to column 0, so `col_indexes[0] = 0`
+  - `values[7] = 80` belongs to column 5, so `col_indexes[7] = 5`
+  - The first entry of row 0 is 10 which is the index 0 in `values`, so `row_indexes[0] = 0`
+  - The first entry of row 1 is 30 which is the index 2 in `values`, so `row_indexes[1] = 2`
+  - The first entry of row 2 is 50 which is the index 4 in `values`, so `row_indexes[2] = 4`
+  - The first entry of row 3 is 80 which is the index 7 in `values`, so `row_indexes[3] = 7`
+  - Row 3 contains only 1 element, so `row_indexes[4] = row_indexes[3] + 1 = 8`
+
+.. Hint::
+
+    We summarize the characteristics of ``row_indexes`` below:
+
+      - It is non-decreasing
+      - Its last entry denotes the number of non-zero entries in the matrix
+      - Its size is `num_rows + 1`
+      - `row_indexes[i+1] - row_indexes[i]` gives the number of non-zero entries in row i
+      - Row `i` contains all zeros if `row_indexes[i+1] == row_indexes[i]`
+
+Now we come back to ragged arrays in k2.
+
+In k2, `row_indexes` is called `row_splits`, compatible with TensorFlow's `RaggedTensor`.
+
+For the following FSA:
+
+.. _ragged1:
+.. figure:: images/ragged.svg
+    :alt: A simple FSA
+    :align: center
+    :figwidth: 600px
+
+    An example FSA.
+
+It is represented in k2 by two arrays:
+
+  - `values = [arc0, arc1, arc2, arc3, arc4, arc5, arc6]`
+  - `row_splits = [0, 3, 4, 6, 7, 7]`
+
+      - Here `arc0` is an instance of C++ class ``Arc``.
+
+`values` saves all of the arcs ordered by state numbers in the FSA.
+
+`row_splits[i]` specifies the where arcs of state `i` begin in the array `values`.
+
+  - `row_splits[0] = 0`, the arcs of state 0 begin at index 0 in `values`
+  - `row_splits[1] = 3`, the arcs of state 1 begin at index 3 in `values`
+  - `row_splits[2] = 4`, the arcs of state 2 begin at index 4 in `values`
+  - `row_splits[3] = 6`, the arcs of state 3 begin at index 6 in `values`
+  - `row_splits` contains 6 entries, so the number of states is `6 - 1 = 5`
+  - The last entry of `row_splits` is 7, so there are 7 arcs in the FSA
+  - `row_splits[1] - row_splits[0] = 3`, so state 0 has 3 arcs
+  - `row_splits[2] - row_splits[1] = 1`, so state 1 has 1 arc
+  - `row_splits[3] - row_splits[2] = 2`, so state 2 has 2 arcs
+  - `row_splits[4] - row_splits[3] = 1`, so state 3 has 1 arc
+  - `row_splits[5] - row_splits[4] = 0`, so state 4 has no arcs at all
+
+**Question**
+  To which state does arc `i` belong?
+
+The above question can be answered in `O(n)` time, where `n` is the number of states,
+by iterating over `row_splits`.
+
+In `k2`, an extra array called `row_ids` is provided to implement such tasks in `O(1)`
+time. For arc `i`, `row_ids[i]` tells the state number to which this arc belongs.
+
+
+.. HINT::
+
+  `row_ids` and `row_splits` contain the same information. `row_ids` is provided to
```

Comment:
[note: nearly the same information.  We can't in general reconstruct the row_splits from the row_ids unless we know the number of lists, because some might be empty.]

Reply:
Thanks, fixed.